src/backend/distributed/planner/local_distributed_join_planner.c

```diff
@@ -0,0 +1,566 @@
+		{
+			*uniqueIndexes = list_append_unique_int(*uniqueIndexes,
+													indexForm->indkey.values[i]);
+		}
+	}
+}
+
+
+/*
+ * RequiredAttrNumbersForRelation returns the required attribute numbers for
+ * the input RTE relation in order for the planning to succeed.
+ *
+ * The function could be optimized by not adding the columns that only appear
+ * WHERE clause as a filter (e.g., not a join clause).
+ */
+static List *
+RequiredAttrNumbersForRelation(RangeTblEntry *relationRte,
+							   RecursivePlanningContext *planningContext)
+{
+	PlannerRestrictionContext *plannerRestrictionContext =
+		planningContext->plannerRestrictionContext;
+
+	/* TODO: Get rid of this hack, find relation restriction information directly */
+	PlannerRestrictionContext *filteredPlannerRestrictionContext =
+		FilterPlannerRestrictionForQuery(plannerRestrictionContext,
+										 WrapRteRelationIntoSubquery(relationRte, NIL));
+
+	RelationRestrictionContext *relationRestrictionContext =
+		filteredPlannerRestrictionContext->relationRestrictionContext;
+	List *filteredRelationRestrictionList =
+		relationRestrictionContext->relationRestrictionList;
+
+	if (list_length(filteredRelationRestrictionList) == 0)
+	{
+		return NIL;
+	}
+	RelationRestriction *relationRestriction =
+		(RelationRestriction *) linitial(filteredRelationRestrictionList);
+
+	PlannerInfo *plannerInfo = relationRestriction->plannerInfo;
+	Query *queryToProcess = plannerInfo->parse;
+	int rteIndex = relationRestriction->index;
+
+	List *allVarsInQuery = pull_vars_of_level((Node *) queryToProcess, 0);
+	ListCell *varCell = NULL;
+
+	List *requiredAttrNumbers = NIL;
+
+	foreach(varCell, allVarsInQuery)
+	{
+		Var *var = (Var *) lfirst(varCell);
+
+		if (var->varno == rteIndex)
+		{
+			requiredAttrNumbers = list_append_unique_int(requiredAttrNumbers,
+														 var->varattno);
+		}
+	}
+
+	return requiredAttrNumbers;
+}
+
+
+/*
+ * CreateRTEToSubqueryConverterContext returns a range table entry which has the most filters
+ * on it along with the restrictions (e.g., fills **restrictionList).
+ *
+ * The function also gets a boolean localTable parameter, so the caller
+ * can choose to run the function for only local tables or distributed tables.
+ */
+static RTEToSubqueryConverterContext *
+CreateRTEToSubqueryConverterContext(RecursivePlanningContext *context,
+									List *rangeTableList)
+{
+	RTEToSubqueryConverterContext *rteToSubqueryConverterContext = palloc0(
+		sizeof(RTEToSubqueryConverterContext));
+
+	int rteIndex = 0;
+	RangeTblEntry *rangeTableEntry = NULL;
+	foreach_ptr(rangeTableEntry, rangeTableList)
+	{
+		rteIndex++;
+		if (rangeTableEntry->rtekind == RTE_SUBQUERY)
+		{
+			rteToSubqueryConverterContext->hasSubqueryRTE = true;
+		}
+
+		/* we're only interested in tables */
+		if (!SubqueryConvertableRelationForJoin(rangeTableEntry))
+		{
+			continue;
+		}
+
+		RTEToSubqueryConverterReference *rteToSubqueryConverter = palloc(
```

Comment:
usually nicer to use palloc0, just in case we add a field and forget to initialize it